scripts/tests/linux/log_line_processing.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2024 Project CHIP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import queue
+import select
+import subprocess
+import threading
+from typing import List
+
+
+class ProcessOutputCapture:
+    """
+    Captures stdout from a process and redirects such stdout to a given file.
+
+    The capture serves several purposes as opposed to just reading stdout:
+      - as data is received, it will get written to a separate file
+      - data is accumulated in memory for later processing (does not starve/block
+        the process stdout)
+      - provides read timeouts for incoming data
+
+    Use as part of a resource mangement block like:
+
+    with ProcessOutputCapture("test.sh", "logs.txt") as p:
+       p.send_to_program("input\n")
+
+       while True:
+           l = p.next_output_line(timeout_sec = 1)
+           if not l:
+               break
+    """
+
+    def __init__(self, command: List[str], output_path: str):
+        # in/out/err are pipes
+        self.command = command
+        self.output_path = output_path
+        self.output_lines = queue.Queue()
+        self.process = None
+        self.io_thread = None
+        self.done = False
+
+    def _io_thread(self):
+        """Reads process lines and writes them to an output file.
+
+        It also sends the output lines to `self.output_lines` for later
+        reading
+        """
+        out_wait = select.poll()
+        out_wait.register(self.process.stdout, select.POLLIN)
+
+        err_wait = select.poll()
+        err_wait.register(self.process.stderr, select.POLLIN)
+
+        with open(self.output_path, "wt") as f:
+            while not self.done:
+                changes = out_wait.poll(0.1)
+                if changes:
+                    out_line = self.process.stdout.readline()
+                    f.write(out_line)
+                    self.output_lines.put(out_line)
+
+                changes = err_wait.poll(0)
+                if changes:
+                    err_line = self.process.stderr.readline()
+                    if err_line:
+                        f.write(f"!!STDERR!! : {err_line}")
```

Comment:
does stderr not go to output_lines?

Reply:
Process exit results in a '' error line. I thought this is ok for output lines, but because I prefix `STDERR!!` in front of errors, I figured empty error lines are annoying (and I did get them while testing) hence this change.

Reply:
ah, ok, gotcha.

Reply:
Removed the logic though ... I think it is probably fine to see a empty stderr at the end.